electron/process_manager.ts

```diff
@@ -67,39 +67,57 @@ const UDP_FORWARDING_TEST_TIMEOUT_MS = 5000;
 const UDP_FORWARDING_TEST_RETRY_INTERVAL_MS = 1000;
 
 // Fulfills iff:
-//  - we can connect to the Shadowsocks server port
 //  - the shadowsocks and tun2socks binaries were started
 //  - the system configured to route all traffic through the proxy
-//  - the remote server has enabled UDP forwarding
-//  - we can speak with a semi-random test site via the proxy
+//  - the connectivity checks pass, if not automatically connecting on startup (e.g. !isAutoConnect)
 //
 // This function is roughly the Electron counterpart of Android's VpnTunnelService.startShadowsocks.
-export function startVpn(config: cordova.plugins.outline.ServerConfig, onDisconnected: () => void) {
-  // RoutingService requires an IP; perform a lookup now and use that throughout setup.
+export function startVpn(
+    config: cordova.plugins.outline.ServerConfig, onDisconnected: () => void,
+    isAutoConnect = false) {
+  return startLocalShadowsocksProxy(config, onDisconnected)
+      .then(() => {
+        if (isAutoConnect) {
+          return;
+        }
+        // Only perform the connectivity checks when we're not automatically connecting on boot,
+        // since we may not have network connectivity.
+        return checkConnectivity(config).then((ip) => {
+          // Cache the resolved IP so it can be stored for auto connect.
+          config.host = ip;
```

Comment:
Sorry, I'm sure I left a comment on this line first time around but it's not there now...

Let's think about how we can improve this: it's not obvious that `startVpn` would re-write its arguments and it's not clear how this results in the ip being cached.

Reply:
Would you like me to add a comment explaining this behavior? Or do you mean improving the approach altogether?

Reply:
To make the behavior explicit, `startVpn` could return `Promise<ServerConfig>` with the resolved hostname.

Reply:
Making the code explicit beats a comment, I think! Let's have it return the new `ServerConfig`.

Reply:
Done in 0f89eb6. Not tested on Windows with the full connection, though I have verified that the logic to return and persist the new config works in macOS.